cmd/controller/apps_reconciler.go

```diff
@@ -35,5 +39,31 @@ func (r *AppsReconciler) Reconcile(request reconcile.Request) (reconcile.Result,
 		return reconcile.Result{}, err
 	}
 
-	return r.appFactory.NewCRDApp(existingApp, log).Reconcile()
+	force := false
+	crdApp := r.appFactory.NewCRDApp(existingApp, log)
+	r.updateAppRefs(crdApp.GetSecretRefs(), "secret", existingApp)
+	r.updateAppRefs(crdApp.GetConfigMapRefs(), "configmap", existingApp)
+	if r.appUpdateStatus.IsUpdateNeeded(existingApp.Name, existingApp.Namespace) {
+		r.appUpdateStatus.MarkUpdated(existingApp.Name, existingApp.Namespace)
+		force = true
+	}
+
+	return crdApp.Reconcile(force)
+}
+
+func (r *AppsReconciler) updateAppRefs(refNames map[string]struct{}, kind string, app *v1alpha1.App) {
+	// If App is being deleted, remove App from appRefTracker.
+	if app.DeletionTimestamp != nil {
+		for refName := range refNames {
+			r.appRefTracker.RemoveAppFromRefMap(kind, refName, app.Namespace, app.Name)
+		}
+	}
+
+	// Make sure refs for App are always up to date
+	// in appRefTracker.
+	for refName := range refNames {
```

Comment:
this i guess doesnt track when refs are deleted. not sure if we need to deal with that...

Reply:
Refs are removed upon deletion via the ConfigMap/Secret handlers. When a Secret/ConfigMap is deleted, the key is removed using the `RemoveRef` func.

Reply:
i meant when ref is not longer used by an app but it still exists within cluster.

Reply:
Okay, makes sense. Maybe what could be done is the AppRefTracker stores `map[refs]map[app]map[kind]ref`. During the updateAppRefs method in apps_reconciler, if the refs don't match, some kind of prune method could be called to update.

Reply:
I added a `PruneAppFromRefs` func to help address this. It requires the `AppRefTracker` to also carry the state what refs an App has.